programs/st_client.c

```diff
@@ -0,0 +1,354 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the project nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.	IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#ifdef _WIN32
+#define _CRT_SECURE_NO_WARNINGS
+#endif
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdarg.h>
+#include <sys/types.h>
+#ifndef _WIN32
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <arpa/inet.h>
+#include <errno.h>
+#include <pthread.h>
+#include <unistd.h>
+#else
+#include <winsock2.h>
+#include <ws2tcpip.h>
+#endif
+#include <usrsctp.h>
+#include "programs_helper.h"
+
+#define MAX_PACKET_SIZE (1<<16)
+#define BUFFER_SIZE 80
+#define DISCARD_PPID 39
+
+#define TIMER_INTERVAL_MSECS 10
+
+static int connecting = 0;
+static int finish = 0;
+
+static unsigned get_tick_count()
+{
+#ifdef _WIN32
+	return GetTickCount();
+#else
+	static const clock_t clocks_per_msec = CLOCKS_PER_SEC / 1000;
+	return clock() / clocks_per_msec;
+#endif
+}
+
+static void
+handle_packets(int sock, struct socket* s, void* sconn_addr)
+{
+	char *dump_buf;
+	ssize_t length;
+	char buf[MAX_PACKET_SIZE];
+
+	fd_set rfds;
+	struct timeval tv;
+
+	unsigned next_fire_time = get_tick_count();
+	unsigned last_fire_time = next_fire_time;
+
+	while (!finish) {
+
+		unsigned now = get_tick_count();
+		int wait_time;
+
+		if ((int) (now - next_fire_time) > 0) {
+			usrsctp_handle_timers(now - last_fire_time);
+			last_fire_time = now;
+			next_fire_time = now + TIMER_INTERVAL_MSECS;
+		}
+
+		wait_time = next_fire_time - now;
+		tv.tv_sec = wait_time / 1000;
+		tv.tv_usec = (wait_time % 1000) * 1000;
+
+		select(1, &rfds, NULL, NULL, &tv);
```

Comment:
Does this really work? `rfds` need to be setup up with
```
FD_ZERO(&rfds);
FD_SET(sock, &rfds);
```
before each `select()` call. You also need to use `sock + 1` as its first argument.